### ISSUE TITLE
Testing code

### ISSUE THREAD

**Issue by ozzyou** [untitled-ai/slot_attention#3]
Hi, thank you very much for the really nice implementation! I have trained the model for 100 epochs and the evaluation results look nice. I was wondering if there's also testing code available. I implemented my own, but I get results such as the image below.

Thank you very much for your reply.

![123](https://user-images.githubusercontent.com/27645665/112996570-8e9bc200-916c-11eb-8b71-ee9a60405ed3.jpg)

**Comment by ozzyou** [comment]
See my implementation below

```
from slot_attention.data import CLEVRDataModule
from slot_attention.method import SlotAttentionMethod
from slot_attention.model import SlotAttentionModel
from slot_attention.params import SlotAttentionParams
from slot_attention.utils import rescale
import torch
from torchvision import transforms
from PIL import Image
from slot_attention.utils import to_rgb_from_tensor
from torchvision import utils as vutils

params = SlotAttentionParams()
model = SlotAttentionModel(
    resolution=params.resolution,
    num_slots=params.num_slots,
    num_iterations=params.num_iterations,
    empty_cache=params.empty_cache,
)
clevr_transforms = transforms.Compose(
    [
        transforms.ToTensor(),
        transforms.Lambda(rescale),  # rescale between -1 and 1
        transforms.Resize(params.resolution),
    ]
)

clevr_datamodule = CLEVRDataModule(
    data_root=params.data_root,
    max_n_objects=params.num_slots - 1,
    train_batch_size=params.batch_size,
    val_batch_size=params.val_batch_size,
    clevr_transforms=clevr_transforms,
    num_train_images=params.num_train_images,
    num_val_images=params.num_val_images,
    num_workers=params.num_workers,
)

root = "/home/ozzy/Projects/slot_attention/"
model = SlotAttentionMethod(model=model, datamodule=clevr_datamodule, params=params)
model.load_state_dict(torch.load(root + "wandb/offline-run-1/files/slot-attention-clevr6/3cy530ay/checkpoints/epoch=99-step=27298.ckpt"), strict=False)
model.eval()

img = Image.open(root + "data/CLEVR_v1.0/images/test/CLEVR_test_014999.png")
img = img.convert("RGB")
img = clevr_transforms(img).unsqueeze(0)

recon_combined, recons, masks, slots = model(img)
out = to_rgb_from_tensor(
    torch.cat(
        [
            img.unsqueeze(1),  # original images
            recon_combined.unsqueeze(1),  # reconstructions
            recons * masks + (1 - masks),  # each slot
        ],
        dim=1,
    )
)
print("RECON SHAPE", out.shape)

batch_size, num_slots, C, H, W = recons.shape
images = vutils.make_grid(
    out.view(batch_size * out.shape[1], C, H, W).cpu(), normalize=False, nrow=out.shape[1],
)

new = transforms.ToPILImage()(images)
new.save("123.jpg")
```

**Comment by brydenfogelman** [comment]
Hmm not sure I quite understand what you're asking for here ... do you just want some code to test the model?

Also, the first image linked is that the results from the model or the test code you pasted?

**Comment by ozzyou** [comment]
Hi, thanks for your reply. Yes, that would be great. The image is the result of the test code.

**Comment by liuyvchi** [comment]
> Hi, thanks for your reply. Yes, that would be great. The image is the result of the test code.

Hi, does your issue been resolved? I meet the same issue as yours.
![image](https://user-images.githubusercontent.com/13773846/119675736-e22a4380-be6f-11eb-854d-5e5823bae6bc.png)

**Comment by greeneggsandyaml** [comment]
I also find that training diverges -- I get similar-looking results (i.e. terrible results) after 100 epochs. 

In other words, we are saying that we cannot get the model to train properly with this code. @brydenfogelman have you been able to successfully train a model with this code?

**Comment by greeneggsandyaml** [comment]
Hello, I just wanted to follow up on this issue.

**Comment by brydenfogelman** [comment]
> I also find that training diverges -- I get similar-looking results (i.e. terrible results) after 100 epochs.
> 
> In other words, we are saying that we cannot get the model to train properly with this code. @brydenfogelman have you been able to successfully train a model with this code?

Hi! I was able to successfully train the model ... the resulting image in the README was from this model. I can try rerunning the model and seeing if I can replicate the issue you all are having.

I may have also introduced a bug in 2fdd3969292dcdc8fb3d6fbc970f9b631053b79e by switching the LR scheduler to match the paper. I'll test this over the weekend.

In the meantime, @greeneggsandyaml you could try reverting the model back to the Exponential LR scheduler and see if that works?

**Comment by lukemelas** [comment]
Hello, thanks for the response! Yes, I will try with the exponential scheduler and report back results here.

**Comment by greeneggsandyaml** [comment]
Hi, thanks for the response! Yes, I will try with the exponential scheduler and report back results here when the experiment is complete.

**Comment by greeneggsandyaml** [comment]
As promised, here is an update. I ran the code from commit `603787ddebde0e19ff9419e6a4e4311ce362956d` and everything worked well! 

For those who are interested, here is my Weights and Biases log: https://wandb.ai/lukemelas2/public-experiments/runs/td2j9zcn?workspace=user-

Overall, this is great to see. I'll be doing more investigation into this as well.

**Comment by greeneggsandyaml** [comment]
Hello, I'm back with another update. Also, @brydenfogelman, did you manage to run the code again? 

I'm finding that sometimes I get results that look good: 
![looks-good](https://user-images.githubusercontent.com/64370631/121198387-002d7600-c86a-11eb-9b03-4c82936da91c.png)

and sometimes I get results that look bad:
![looks-bad](https://user-images.githubusercontent.com/64370631/121198382-fefc4900-c869-11eb-9876-4462c434b4ef.png)

Have you seen these sorts of "splotchy" results before? Is it just due random initialization? It feels to me like it is too much variation to be caused solely by random initialization.

**Comment by brydenfogelman** [comment]
@greeneggsandyaml How long did you train for? 

I think even the original authors found that results can vary on the network and slot initialization. I think this figure from the paper demonstrates this finding.
![image](https://user-images.githubusercontent.com/9598877/121420201-a5efec00-c921-11eb-9942-925f555fff1f.png)

(Looking at this caption again I also realized that I didn't increase the number of slot iteration at test time, increasing this would probably make the results look better)

Here's an image of one of my earlier experiments where it did randomly learn to separate the background image.
![image](https://user-images.githubusercontent.com/9598877/121419935-5f01f680-c921-11eb-849a-8ecad969eb90.png)

They also trained their model for significantly longer than I trained it here (5 days wall clock time).

My best guess is that increasing the number of slot iterations at test time will improve the visualizations. What are your thoughts @greeneggsandyaml?

**Comment by ZiwenZhuang** [comment]
> Hello, I'm back with another update. Also, @brydenfogelman, did you manage to run the code again?
> 
> I'm finding that sometimes I get results that look good:
> ![looks-good](https://user-images.githubusercontent.com/64370631/121198387-002d7600-c86a-11eb-9b03-4c82936da91c.png)
> 
> and sometimes I get results that look bad:
> ![looks-bad](https://user-images.githubusercontent.com/64370631/121198382-fefc4900-c869-11eb-9876-4462c434b4ef.png)
> 
> Have you seen these sorts of "splotchy" results before? Is it just due random initialization? It feels to me like it is too much variation to be caused solely by random initialization.

Hi, How did you solve the problem?

I used the test code from above but get the gray-scale output.
![123](https://user-images.githubusercontent.com/28809433/130380023-c2fda425-c02f-4410-a3c9-c14790ae67d7.jpg)

I checked the images in wandb log files, which looks acceptable.
![images_13699_0](https://user-images.githubusercontent.com/28809433/130380083-3c182a5a-c235-4644-b2d6-60dd69c4ba4b.png)

Is there anything wrong in the testing code above? How should I change it?

Thank you,

**Comment by ZiwenZhuang** [comment]
Hi, this is a follow-up, I know where the test code above might go wrong.

Depending on how you save your checkpoint file, the state_dict loaded might not match the model state_dict. Using `strict = False` might failed to load any weights into the model.

You should check the state_dict keys while loading the model.

:)

**Comment by brydenfogelman** [comment]
@ZiwenZhuang mentioned above is that the LR scheduler changed with ended up breaking subsequent runs. I'll try and push the fixes to the repo later today
